### PR TITLE
Add pod level metrics for edge node

### PR DESCRIPTION
**Pull request by zhu733756**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
add pod level metrics for edge node

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for reviewers**:
```
```

**Additional documentation, usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by benjaminhuo** [comment]
@zhu733756 it's better to merge these 2 commit into one and force push again

**Comment by zhu733756** [comment]
> @zhu733756 it's better to merge these 2 commit into one and force push again

@benjaminhuo Ok, Node metrics also need to modify.

**Comment by benjaminhuo** [comment]
> > @zhu733756 it's better to merge these 2 commit into one and force push again
> 
> @benjaminhuo Ok, Node metrics also need to modify.

Node metrics is another PR

**Comment by benjaminhuo** [comment]
[under developing] should be [WIP]

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 1]
```diff
@@ -319,118 +509,217 @@ func (m metricsServer) GetNamedMetricsOverTime(metrics []string, start, end time
...
+		return m.GetLevelNodeNamedMetricsOverTime(metrics, start, end, step, opts)
+	case monitoring.LevelPod:
+		return m.GetLevelPodNamedMetricsOverTime(metrics, start, end, step, opts)
+	default:
+		return res
+	}
+
+}
+
+func (m metricsServer) GetLevelNodeNamedMetricsOverTime(metrics []string, start, end time.Time, step time.Duration, opts *monitoring.QueryOptions) []monitoring.Metric {
```
Comment (benjaminhuo) [inline pkg/simple/client/monitoring/metricsserver/metricsserver.go:529]: GetLevelNodeNamedMetricsOverTime => GetNodeLevelNamedMetricsOverTime

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 2]
```diff
@@ -319,118 +509,217 @@ func (m metricsServer) GetNamedMetricsOverTime(metrics []string, start, end time
...
+		if ok {
+			res = append(res, monitoring.Metric{MetricName: enm, MetricData: *nodeMetrics[enm]})
+		}
+	}
+	return res
+}
 
-			m.Usage.DeepCopyInto(&usage)
-			status[m.Name].Capacity.DeepCopyInto(&cap)
+func (m metricsServer) GetLevelPodNamedMetricsOverTime(metrics []string, start, end time.Time, step time.Duration, opts *monitoring.QueryOptions) []monitoring.Metric {
```
Comment (benjaminhuo) [inline pkg/simple/client/monitoring/metricsserver/metricsserver.go:638]: GetLevelPodNamedMetricsOverTime => GetPodLevelNamedMetricsOverTime

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 3]
```diff
@@ -144,14 +145,21 @@ func (m metricsServer) getNodeMetricsFromMetricsAPI() (*metricsapi.NodeMetricsLi
 }
 
 // pods metrics of edge nodes
-func (m metricsServer) getPodMetricsFromMetricsAPI(ns string, podName string) ([]metricsapi.PodMetrics, error) {
-	if ns == "" {
-		ns = metav1.NamespaceAll
+func (m metricsServer) getPodMetricsFromMetricsAPI(edgeNodePodsFiltered map[string]bool, ns string) ([]metricsapi.PodMetrics, error) {
+
+	var isNamespaceInEdgeNodePodsFilteredKey bool
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 4]
```diff
@@ -103,12 +106,33 @@ func (m metricsServer) filterEdgeNodeNames(edgeNodes map[string]v1.Node, opts *m
 	return edgeNodeNamesFiltered
 }
 
+func (m metricsServer) filterEdgeNodePodNames(opts *monitoring.QueryOptions) map[string]bool {
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 5]
```diff
@@ -103,12 +106,33 @@ func (m metricsServer) filterEdgeNodeNames(edgeNodes map[string]v1.Node, opts *m
...
+
+	edgeNodePodsFiltered := make(map[string]bool)
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 6]
```diff
@@ -120,6 +144,65 @@ func (m metricsServer) getNodeMetricsFromMetricsAPI() (*metricsapi.NodeMetricsLi
 	return metrics, nil
 }
 
+// pods metrics of edge nodes
+func (m metricsServer) getPodMetricsFromMetricsAPI(edgeNodePodsFiltered map[string]bool, ns string) ([]metricsapi.PodMetrics, bool, error) {
+
+	var isNamespaceInEdgeNodePodNames bool
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 7]
```diff
@@ -120,6 +144,65 @@ func (m metricsServer) getNodeMetricsFromMetricsAPI() (*metricsapi.NodeMetricsLi
...
+
+	for p, _ := range edgeNodePodsFiltered {
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 8]
```diff
@@ -183,6 +267,20 @@ const (
 
 var edgeNodeMetrics = []string{metricsNodeCPUUsage, metricsNodeCPUTotal, metricsNodeCPUUltilisation, metricsNodeMemoryUsageWoCache, metricsNodeMemoryTotal, metricsNodeMemoryUltilisation}
 
+// pod metrics definition
+const (
+	metricsPodCPUUsage           = "pod_cpu_usage"
+	metricsPodMemoryUsageWoCache = "pod_memory_usage_wo_cache"
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 9]
```diff
@@ -199,118 +297,210 @@ func (m metricsServer) GetNamedMetrics(metrics []string, ts time.Time, o monitor
...
+	default:
+		return res
+	}
+}
 
-		edgeNodeNamesFiltered := m.filterEdgeNodeNames(edgeNodes, opts)
-		if len(edgeNodeNamesFiltered) == 0 {
-			klog.V(4).Infof("No edge node metrics is requested")
-			return res
+func (m metricsServer) GetLevelNodeNamedMetrics(metrics []string, ts time.Time, opts *monitoring.QueryOptions) []monitoring.Metric {
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 10]
```diff
@@ -199,118 +297,210 @@ func (m metricsServer) GetNamedMetrics(metrics []string, ts time.Time, o monitor
...
+		if ok {
+			res = append(res, monitoring.Metric{MetricName: enm, MetricData: *nodeMetrics[enm]})
+		}
+	}
+	return res
+}
 
-			m.Usage.DeepCopyInto(&usage)
-			status[m.Name].Capacity.DeepCopyInto(&cap)
+func (m metricsServer) GetLevelPodNamedMetrics(metrics []string, ts time.Time, opts *monitoring.QueryOptions) []monitoring.Metric {
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go (hunk 1 above)
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go (hunk 2 above)
Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 11]
```diff
@@ -194,24 +193,29 @@ func (m metricsServer) getPodMetricsFromMetricsAPI(edgePods map[string]bool, pod
 			}
 			combinedPodMetrics = append(combinedPodMetrics, *metrics)
 		}
-		return combinedPodMetrics, isPodMetricsUsingListRequst, nil
+		return combinedPodMetrics, nil
 	}
 
 	// use list request in anthor cases
```
Comment (benjaminhuo) [inline pkg/simple/client/monitoring/metricsserver/metricsserver.go:199]: use list request in anthor cases
=>
use list request for other cases

Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

**Review by benjaminhuo — APPROVED** [review]
/lgtm

File: pkg/simple/client/monitoring/metricsserver/metricsserver.go [hunk 12]
```diff
@@ -319,118 +520,212 @@ func (m metricsServer) GetNamedMetricsOverTime(metrics []string, start, end time
...
+					metricValues[enm].Metadata["host_ip"] = addr.Address
+				}
+				break
+			}
 		}
 
-		status := make(map[string]v1.NodeStatus)
-		for n, _ := range edgeNodeNamesFiltered {
-			status[n] = edgeNodes[n].Status
+		_, ok = metricsMap[metricsNodeCPUUsage]
```
Comment (zryfish) [inline pkg/simple/client/monitoring/metricsserver/metricsserver.go:607]: DRY, I think we can refactor these code with for range, for example
```
for k, v := range metricsMap {
  switch k {
  case metricsNodeCPUUsage:
      .....
  case metricsNodeCPUTotal:
    ....
  }
}
```

Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: fixed

Comment (zryfish) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: `range` a `map` returns `key, value`, the second returned value is not `ok`,  checkout this example https://gobyexample.com/range

Comment (zhu733756) [reply pkg/simple/client/monitoring/metricsserver/metricsserver.go]: The  type of  the values in the map here is bool .  Maybe `ok` in golang has special meaning.  I think I can modify it to improve readability.

**Comment by zryfish** [comment]
@zhu733756 conflicts

**Comment by zhu733756** [comment]
> @zhu733756 conflicts

**Comment by zryfish** [comment]
Please squash commits, the rest is ok.

**Comment by zryfish** [comment]
conflicts again

**Comment by zhu733756** [comment]
Managed, it doesn't look very clear, Shall we post a new pr?

**Comment by zryfish** [comment]
> Managed, it doesn't look very clear, Shall we post a new pr?

What do you mean?

**Comment by benjaminhuo** [comment]
> Managed, it doesn't look very clear, Shall we post a new pr?

This PR is OK as long as the code conflicts are resolved, why do we need a new PR for this?

**Comment by zhu733756** [comment]
> > Managed, it doesn't look very clear, Shall we post a new pr?
> 
> What do you mean?

I can't find a proper way to manage two commits to one, because it confilicts.

**Comment by benjaminhuo** [comment]
> > > Managed, it doesn't look very clear, Shall we post a new pr?
> > 
> > 
> > What do you mean?
> 
> I can't find a proper way to manage two commits to one, because it confilicts, I have to do a managed logic.

rebase could resolve this

**Comment by zhu733756** [comment]
> > > > Managed, it doesn't look very clear, Shall we post a new pr?
> > > 
> > > 
> > > What do you mean?
> > 
> > 
> > I can't find a proper way to manage two commits to one, because it confilicts, I have to do a managed logic.
> 
> rebase could resolve this

fixed

**Pull request merged**